### PR TITLE
analyze: add "archs" verb that lists all known architectures

### PR DESCRIPTION
**Pull request by poettering**

### CODE REVIEW

**Comment by poettering** [comment]
This is a minor addition, but see no reason to sneak it into v255. Hence let's leave it for v256

**Review by bluca — COMMENTED** [review]
Please add a call to this in TEST-65-ANALYZE - no need to verify its output, but just so that there's some coverage from sanitizer et al

File: src/analyze/analyze-archs.c [hunk 1]
```diff
@@ -0,0 +1,89 @@
+/* SPDX-License-Identifier: LGPL-2.1-or-later */
+
+#include "analyze.h"
+#include "analyze-archs.h"
+#include "format-table.h"
+
+static int add_arch(Table *t, Architecture a) {
+        const char *c = NULL, *color = NULL;
+        int r;
+
```
Comment (YHNdnzj) [inline src/analyze/analyze-archs.c]: Missing assertion for `t`

File: src/analyze/analyze-archs.c [hunk 2]
```diff
@@ -0,0 +1,89 @@
+/* SPDX-License-Identifier: LGPL-2.1-or-later */
+
+#include "analyze.h"
+#include "analyze-archs.h"
+#include "format-table.h"
+
+static int add_arch(Table *t, Architecture a) {
+        const char *c = NULL, *color = NULL;
```
Comment (YHNdnzj) [inline src/analyze/analyze-archs.c]: No need to initialize these?

File: src/analyze/analyze-archs.c [hunk 3]
```diff
@@ -0,0 +1,89 @@
...
+                                a = architecture_from_string(argv[i]);
+                        if (a < 0)
+                                return log_error_errno(a, "Architecture \"%s\" not known.", argv[i]);
+
+                        r = add_arch(table, a);
+                        if (r < 0)
+                                return r;
+                }
+
+                 (void) table_set_sort(table, (size_t) 0);
```
Comment (YHNdnzj) [inline src/analyze/analyze-archs.c]:
```suggestion
                (void) table_set_sort(table, (size_t) 0);
```

File: src/analyze/analyze-archs.c [hunk 4]
```diff
@@ -0,0 +1,89 @@
...
+
+        if (strv_isempty(strv_skip(argv, 1)))
+
+                for (Architecture a = 0; a < _ARCHITECTURE_MAX; a++) {
+                        r = add_arch(table, a);
+                        if (r < 0)
+                                return r;
+                }
+        else {
+                for (int i = 1; i < argc; i++) {
```
Comment (yuwata) [inline src/analyze/analyze-archs.c]: `STRV_FOREACH(x, strv_skip(argv, 1))`

File: test/units/testsuite-65.sh [hunk 5]
```diff
@@ -866,6 +866,8 @@ systemd-analyze pcrs
 systemd-analyze pcrs --json=pretty
 systemd-analyze pcrs 14 7 0 ima
 
+systemd-analyze archs
```
Comment (yuwata) [inline test/units/testsuite-65.sh]: Please also call with several names.

**Comment by poettering** [comment]
Addressed all issues raised.

**Review by bluca — APPROVED** [review]

**Review by YHNdnzj — APPROVED** [review]

**Comment by keszybz** [comment]
Hmm, shouldn't this support `--json`?

Also, is it "archs" or "arches"? "archs" looks strange and google says that it's not the correct spelling. That that is for "arch" in the sense of an architectural element, not our jargon abbreviation for "architecture". But we have "es" in "architectures", so I think we should here too.

**Comment by poettering** [comment]
To avoid the archs vs. arches thing i renamed this to `architectures` now. No need to abbreviate here I think, it's not the most common of commands to execute I guess.

I also added the JSON output, and while at it also added support for it to "capability" and "exit-status" because trivial

**Comment by bluca** [comment]
Please also add shell completion at least for bash

File: src/analyze/analyze.c [hunk 6]
```diff
@@ -14,6 +14,7 @@
 #include "alloc-util.h"
 #include "analyze.h"
 #include "analyze-blame.h"
+#include "analyze-architectures.h"
```
Comment (yuwata) [inline src/analyze/analyze.c:16]: Not sorted.

**Review by yuwata — APPROVED** [review]

**Comment by poettering** [comment]
> Please also add shell completion at least for bash

happy to review/merge a patch for that, but I don't think i am a bash guru enough to write that.

**Comment by bluca** [comment]
> > Please also add shell completion at least for bash
> 
> happy to review/merge a patch for that, but I don't think i am a bash guru enough to write that.

It doesn't need to be fancy, just list the verb, simply add it to the STANDALONE list at  https://github.com/systemd/systemd/blob/main/shell-completion/bash/systemd-analyze#L65

**Comment by poettering** [comment]
> It doesn't need to be fancy, just list the verb, simply add it to the STANDALONE list at https://github.com/systemd/systemd/blob/main/shell-completion/bash/systemd-analyze#L65

That doesn't look right nor helpful. The command does take parameters after all.

File: src/analyze/analyze.c (hunk 6 above)
Comment (poettering) [reply src/analyze/analyze.c:16]: fixed.

**Pull request merged**